database/mysql.go

```diff
@@ -94,30 +99,69 @@ func (my *mysql) Connect(ctx context.Context) (*sql.DB, error) {
 	return db, nil
 }
 
-func mysqlConnection(logger log.Logger, user, pass string, address string, database string, sslCA string) *mysql {
+func mysqlConnection(logger log.Logger, mysqlConfig *MySQLConfig, databaseName string) (*mysql, error) {
 	timeout := "30s"
 	if v := os.Getenv("MYSQL_TIMEOUT"); v != "" {
 		timeout = v
 	}
+	params := fmt.Sprintf(`timeout=%s&charset=utf8mb4&parseTime=true&sql_mode="ALLOW_INVALID_DATES,STRICT_ALL_TABLES"`, timeout)
 
-	var params []string
-	params = append(params, fmt.Sprintf("timeout=%s", timeout))
-	params = append(params, "charset=utf8mb4")
-	params = append(params, "parseTime=true")
-	params = append(params, `sql_mode="ALLOW_INVALID_DATES,STRICT_ALL_TABLES"`)
+	var tlsConfig *tls.Config
 
-	if sslCA != "" {
-		params = append(params, fmt.Sprintf("sslca=%s", sslCA))
+	if mysqlConfig.UseTLS {
+		// If any custom options are set then we need to create a custom TLS configuration. Otherwise we can just set
+		// tls=true in the DSN
+		if mysqlConfig.InsecureSkipVerify || mysqlConfig.TLSCAFile != "" || mysqlConfig.VerifyCAFile {
+			tlsConfig = &tls.Config{
+				InsecureSkipVerify: mysqlConfig.InsecureSkipVerify,
+			}
+
+			if mysqlConfig.TLSCAFile != "" {
+				rootCertPool := x509.NewCertPool()
+				certPem, err := ioutil.ReadFile(mysqlConfig.TLSCAFile)
+				if err != nil {
+					return nil, err
+				}
+
+				block, _ := pem.Decode(certPem)
+				caCert, err := x509.ParseCertificate(block.Bytes)
+				if err != nil {
+					return nil, err
+				}
+
+				if appendOK := rootCertPool.AppendCertsFromPEM(certPem); !appendOK {
+					return nil, err
+				}
+				tlsConfig.RootCAs = rootCertPool
+
+				if mysqlConfig.VerifyCAFile {
+					tlsConfig.VerifyConnection = func(state tls.ConnectionState) error {
+						if !(state.PeerCertificates[len(state.PeerCertificates)-1].Equal(caCert)) {
```

Comment:
We should perform a length check here, right? This could panic if the chain has one certificate.

Reply:
It'd panic if it had 0 certificates (length 1 - 1 = 0 which is a one-element slice) and [the docs](https://pkg.go.dev/crypto/tls) state

> // On the client side, it can't be empty.